fish_completions.go

```diff
@@ -30,7 +30,7 @@ end
 function __%[1]s_perform_completion
     __%[1]s_debug "Starting __%[1]s_perform_completion"
 
-    set args (string split -- " " (commandline -c))
+    set args (string split -- " " (string trim -l (commandline -c)))
```

Comment:
Should be `string trim -- -l (commandline -c)`

---

Often it's not necessary to pass the current token to fish.
Rather, one can just compute all possible tokens because fish will filter them anyway.
Sometimes it can be useful to do the filtering in the app (for example if it's not feasible to enumerate all possible tokens), so it's fine.

Reply:
> Should be `string trim -- -l (commandline -cp)`

I don't understand from the documentation what the `-p` flag does.  Could you clarify?

> Often it's not necessary to pass the current token to fish.
> Rather, one can just compute all possible tokens because fish will filter them anyway.
> Sometimes it can be useful to do the filtering in the app (for example if it's not feasible to enumerate all possible tokens), so it's fine. (the `testprog` examples show that the current token is necessary to complete `--flags`)

Right.  So originally, I thought the program should do the filtering, but with experience, I now realize it is actually better to provide `fish` (and `zsh`) with all unfiltered completions so that the shell can do smarter matching than simply on prefix.  For example, if the user types `prog paw<TAB>`, giving shell completions like `bearpaw` will allow it to still match even though it is not on the prefix.

As you saw in this PR however, I didn't find a way to find out if `fish` had found a smart match or not, so, when wanting to do file completion, I was forced to filter on prefix in the script itself, to know if there would be a match or not.  This is because Cobra only wants to do file completion if there are no valid completions, so it can be consistent with `bash` and `zsh`.  This was not ideal, so if you know of a way to figure out in the script if fish will match on a list of completion, that would be helpful.

Reply:
The `-p` is explained in https://github.com/spf13/cobra/pull/1249#discussion_r563333005. Sorry for the wall of text :smile_cat: 

fish does fuzzy substring matching if there is no prefix or exact match. This is currently not exposed in a scriptable way.

I wonder if it's acceptable to do file completions only if there are no results. Doesn't sound too good, so probably the prefix comparison is better (maybe make that substring matching instead of only prefix).

Reply:
> I wonder if it's acceptable to do file completions only if there are no results. 

You mean if the program returns no completions?  This is what the script does right now (before this PR). But that requires that the program do the filtering on itself.  This PR is (in part) trying to improve that situation. By doing the filtering in the fish script, it will allow the program to return all completions, which will then allow the fuzzy matching you mention.  The fuzzy matching will only not work when file completion is requested or when nospace is requested, both of which are the less common cases.

> Doesn't sound too good, so probably the prefix comparison is better (maybe make that substring matching instead of only prefix).

As long as we use an algorithm that is stricter than what fish does, we could indeed improve compared to filtering on prefix.  Do you know where I could see the algorithm used by fish?  Or else I'll just play around and try to see if any pattern match seems to work.

Reply:
Just to clarify for myself; the common cases are
- no files
- only files because the program returns no completions

in those cases fish already does all the filtering. But if we're in between - the program returns some completions but they may not match the token - we want to filter in the completion script. That sounds great.

I agree that this is probably not very common. Most programs should take either custom tokens (like subcommands) or files, not both. Git is a notable exception, but Git completion computes custom file paths based on `git ls-files`.

I'd just ignore nospace for fish for now until, it behaves the right way most of the time already.
I can't think of a good solution right now, it needs support from fish.

> Do you know where I could see the algorithm used by fish?

The two most permissive filters are [here](https://github.com/fish-shell/fish-shell/blob/062f24d91b199780ae7a8493988549a5f651423f/src/wcstringutil.cpp#L206-L214). I think they can be implemented like this:

```fish
$ function fishy_match -a needle haystack
      if string match -irq -- (string escape --style=regex -- $needle) $haystack
          echo substr match
      end
      if string match -rq -- (string split -- '' $needle | string escape --style=regex | string join '.*?') $haystack
          echo subseq match
      end
  end
$ fishy_match String 'case-insensitive substring match'
substr match
$ fishy_match foo follow
subseq match
```

I use `-r` and regex escaping because the needle might contain globs, which would be interpreted by `string match` without `-r`.

Reply:
> Most programs should take either custom tokens (like subcommands) or files, not both. 

There are some cases where both are valid.  Helm for example can accept the reference to a remote chart, or the local path to a chart.  What I've done in helm is have the program return the list of files as completions to do both file completion and custom token completion.

I think it is important to keep all shells behaving the same for Cobra because the program returns a list of completions to be used by all shells; we don't want to make the program have to behave differently for each shell.

But we could, as a separate effort, support doing file completion even if there are custom token completion matches.  This was not Cobra's previous behaviour, so I had never considered it myself, but I believe it could be done for all shells.  Maybe a future effort...

> I'd just ignore nospace for fish for now until, it behaves the right way most of the time already.
> I can't think of a good solution right now, it needs support from fish.

Well, our current solution is good until fish 3.2.  And then, when fish shows the choices including the fake one with a trailing `.`,  it may be confusing, but I think it may be not much worse than letting the shell add and unwanted space.

PENDING - I'll turn off adding the extra `.` when fish handles the case itself following the characters you mentioned.

PENDING - And I'll also adapt the filtering based on the algos you've provided.

I feel that after all that, we'll have something really solid for fish completion.

Reply:
> What I've done in helm is have the program return the list of files as completions to do both file completion and custom token completion.

Ideally we could tell the shell that we also want file completons, because it can see through variables to tokens like `$PWD/`. Something for the future I guess. The implementation would be a bit hacky:
essentially, you'd need two calls to `complete -c`, one with `-f` and one without.
Each of them has a condition (the `-n` part). If the condition for the one with file completions is true, then fish will also provide file completions (in addition to the provided custom completions).

> we don't want to make the program have to behave differently for each shell.

In terms of what completions to offer, definitely. Of course shells differ in how they display completions.
By default, fish sorts completions, but you can add the `--keep-order` flag to the `complete -c` call, then the program is in control of the ordering.

> I feel that after all that, we'll have something really solid for fish completion.

Yeah, latest changes look great! For some programs  it's hard to compete with manually written completions, but it should work great for conventional programs.